pkg/cmd/experimental/syncgroups/sync-groups.go

```diff
@@ -0,0 +1,174 @@
+
+	// Stderr is the writer to write warnings and errors to
+	Stderr io.Writer
+
+	// Out is the writer to write output to
+	Out io.Writer
+
+	// Quiet controls whether warnings are output
+	Quiet bool
+}
+
+func NewSyncGroupsOptions() *SyncGroupsOptions {
+	return &SyncGroupsOptions{
+		Stderr:           os.Stderr,
+		WhitelistSources: util.StringList{},
+	}
+}
+
+func NewCmdSyncGroups(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := NewSyncGroupsOptions()
+	options.Out = out
+
+	cmd := &cobra.Command{
+		Use:     fmt.Sprintf("%s TYPE [WHITELIST] CONFIG-SOURCE", name),
+		Short:   "Sync OpenShift Groups with records from an external provider.",
+		Long:    syncGroupsLong,
+		Example: fmt.Sprintf(syncGroupsExamples, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			if err := options.Complete(args, f); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			if err := options.Validate(); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(c, err.Error()))
+			}
+
+			syncer, err := options.CreateSyncer()
+			cmdutil.CheckErr(err)
+
+			errs := syncer.Sync()
+			cmdutil.CheckErr(kerrs.NewAggregate(errs))
+		},
+	}
+
+	cmd.Flags().BoolVarP(&options.Quiet, "quiet", "q", options.Quiet, "Suppress warnings")
+	cmd.Flags().StringVar(&options.ConfigSource, "sync-config", "", "The path to the sync config")
+	cmdutil.AddPrinterFlags(cmd)
+
+	return cmd
+}
+
+func (o *SyncGroupsOptions) Complete(args []string, f *clientcmd.Factory) error {
+	// Fill type from args[0]
+	if len(args) > 0 {
+		o.Type = SyncType(args[0])
+	}
+
+	// Add sources from args[1:...]
+	if len(args) > 1 {
+		o.WhitelistSources = append(o.WhitelistSources, args[1:]...)
+	}
+
+	if f != nil {
+		osClient, _, err := f.Clients()
+		if err != nil {
+			return err
+		}
+		o.GroupInterface = osClient.Groups()
+	}
+
+	return nil
+}
+
+func (o *SyncGroupsOptions) Validate() error {
+	if len(o.Type) == 0 {
+		return errors.New("sync type is required")
+	}
+	switch o.Type {
+	case SyncTypeLDAPAll:
+	case SyncTypeLDAPWhitelist:
+	case SyncTypeOpenShiftAll:
+	case SyncTypeOpenShift:
+	default:
+		return fmt.Errorf("sync type must be one of the following: %v", []SyncType{SyncTypeLDAPAll, SyncTypeLDAPWhitelist, SyncTypeOpenShift, SyncTypeOpenShiftAll})
+	}
+	if len(o.ConfigSource) == 0 {
+		return errors.New("sync config source is required")
+	}
+	if (o.Type == SyncTypeLDAPWhitelist || o.Type == SyncTypeOpenShift) && len(o.WhitelistSources.String()) == 0 {
+		return fmt.Errorf("a path to a whitelist file or a list of unique group identifiers is required for sync type %s", o.Type)
+	}
+	return nil
+}
+
+func (o *SyncGroupsOptions) CreateSyncer() (GroupSyncer, error) {
+	// load config file
+	// start building options
+	// add:
+	// ---groupclient for current context
+	// ---groupsyncsource
+	// ---groupsyncscope
+	// ------if it's not an -all, inject whitelist
+	// create syncer
+	return nil, nil
```

Comment:
Build the syncer right here.  You should be able to take your options, build the intermediate object directly in this method and compose your syncer in place right here.  Completely delete your existing `LDAPSyncCommonConfig` and anythign else I marked up there.